README.rst
=============================
Django Channels PubSub
=============================

.. image:: https://badge.fury.io/py/dj-channels-pubsub.svg
    :target: https://badge.fury.io/py/dj-channels-pubsub

.. image:: https://travis-ci.org/bgervan/dj-channels-pubsub.svg?branch=master
    :target: https://travis-ci.org/bgervan/dj-channels-pubsub

.. image:: https://codecov.io/gh/bgervan/dj-channels-pubsub/branch/master/graph/badge.svg
    :target: https://codecov.io/gh/bgervan/dj-channels-pubsub

Django Channels Layer for Google PubSub

Documentation
-------------

The full documentation is at https://dj-channels-pubsub.readthedocs.io.

Quickstart
----------

Install Django Channels PubSub::

    pip install dj-channels-pubsub


Add Environment variables:

::

    export GOOGLE_CLOUD_PROJECT="" # Real project ID required for local emulator too


For the local PubSub Emulator, add:

::

    export PUBSUB_EMULATOR_HOST=localhost:8085

For the local PubSub Emulator with Docker Compose:

::

    export PUBSUB_EMULATOR_HOST=pubsub:8085

::

    pubsub:
    image: google/cloud-sdk:331.0.0
    command: gcloud beta emulators pubsub start --host-port=0.0.0.0:8085
    ports:
      - "8085:8085"
    restart: unless-stopped

Features
--------

* TODO

Running Tests
-------------

Does the code actually work?

::

    source <YOURVIRTUALENV>/bin/activate
    (myenv) $ pip install tox
    (myenv) $ tox


Development commands
---------------------

::

    pip install -r requirements_dev.txt
    invoke -l


Credits
-------

Tools used in rendering this package:

*  Cookiecutter_
*  `cookiecutter-djangopackage`_

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`cookiecutter-djangopackage`: https://github.com/pydanny/cookiecutter-djangopackage


HISTORY.rst
.. :changelog:

History
-------

0.1.0 (2021-08-04)
++++++++++++++++++

* First release on PyPI.


tests/urls.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals, absolute_import

from django.conf.urls import url, include
from django.contrib import admin

urlpatterns = [
    url(r'^admin/', admin.site.urls),
    url(r'^', include('dj_channels_pubsub.urls', namespace='dj_channels_pubsub')),
]


tests/settings.py
"""
Django settings for example project.

Generated by Cookiecutter Django Package
"""

import os

# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/1.9/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = "px@w($hn59hpk*_2zz9l&)=3x0zgg=(zg1mu8v)fp(w*u@438l"

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = True

ALLOWED_HOSTS = []

# Application definition

INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',

    'dj_channels_pubsub',
    

    # if your app has other dependencies that need to be added to the site
    # they should be added here
]

MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
]

ROOT_URLCONF = 'tests.urls'

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [os.path.join(BASE_DIR, 'templates'), ],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        },
    },
]

WSGI_APPLICATION = 'tests.wsgi.application'

# Database
# https://docs.djangoproject.com/en/1.9/ref/settings/#databases

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': os.path.join(BASE_DIR, 'db.sqlite3'),
    }
}

# Password validation
# https://docs.djangoproject.com/en/1.9/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
    },
]

# Internationalization
# https://docs.djangoproject.com/en/1.9/topics/i18n/

LANGUAGE_CODE = 'en-us'

TIME_ZONE = 'UTC'

USE_I18N = True

USE_L10N = True

USE_TZ = True

# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/1.9/howto/static-files/

STATIC_URL = '/static/'


tests/test_models.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_dj-channels-pubsub
------------

Tests for `dj-channels-pubsub` models module.
"""

from django.test import TestCase

from dj_channels_pubsub import models


class TestDj_channels_pubsub(TestCase):

    def setUp(self):
        pass

    def test_something(self):
        pass

    def tearDown(self):
        pass


tasks.py
import os
import webbrowser

from invoke import task


def open_browser(path):
    try:
        from urllib import pathname2url
    except:
        from urllib.request import pathname2url
    webbrowser.open("file://" + pathname2url(os.path.abspath(path)))


@task
def clean_build(c):
    """
    Remove build artifacts
    """
    c.run("rm -fr build/")
    c.run("rm -fr dist/")
    c.run("rm -fr *.egg-info")


@task
def clean_pyc(c):
    """
    Remove python file artifacts
    """
    c.run("find . -name '*.pyc' -exec rm -f {} +")
    c.run("find . -name '*.pyo' -exec rm -f {} +")
    c.run("find . -name '*~' -exec rm -f {} +")


@task
def coverage(c):
    """
    check code coverage quickly with the default Python
    """
    c.run("coverage run --source dj-channels-pubsub runtests.py tests")
    c.run("coverage report -m")
    c.run("coverage html")
    c.run("open htmlcov/index.html")


@task
def docs(c):
    """
    Build the documentation and open it in the browser
    """
    c.run("rm -f docs/dj-channels-pubsub.rst")
    c.run("rm -f docs/modules.rst")
    c.run("sphinx-apidoc -o docs/ dj_channels_pubsub")

    c.run("sphinx-build -E -b html docs docs/_build")
    open_browser(path='docs/_build/html/index.html')


@task
def test_all(c):
    """
    Run tests on every python version with tox
    """
    c.run("tox")


@task
def clean(c):
    """
    Remove python file and build artifacts
    """
    clean_build(c)
    clean_pyc(c)


@task
def unittest(c):
    """
    Run unittests
    """
    c.run("python manage.py test")


@task
def lint(c):
    """
    Check style with flake8
    """
    c.run("flake8 dj-channels-pubsub tests")


@task(help={'bumpsize': 'Bump either for a "feature" or "breaking" change'})
def release(c, bumpsize=''):
    """
    Package and upload a release
    """
    clean(c)
    if bumpsize:
        bumpsize = '--' + bumpsize

    c.run("bumpversion {bump} --no-input".format(bump=bumpsize))

    import dj_channels_pubsub
    c.run("python setup.py sdist bdist_wheel")
    c.run("twine upload dist/*")

    c.run('git tag -a {version} -m "New version: {version}"'.format(version=dj_channels_pubsub.__version__))
    c.run("git push --tags")
    c.run("git push origin master")


dj_channels_pubsub/layers.py
import asyncio
import functools
import os
import random
import string
import time
import uuid

from channels.exceptions import ChannelFull
from channels.layers import BaseChannelLayer

from google.cloud import pubsub
from .utils import SetQueue


class GcePubSubChannelLayer(BaseChannelLayer):
    """
    Google cloud Pub/Sub channel layer implementation
    """

    def __init__(
        self,
        expiry=60,
        group_expiry=86400,
        capacity=100,
        channel_capacity=None,
        **kwargs
    ):
        super().__init__(
            expiry=expiry,
            capacity=capacity,
            channel_capacity=channel_capacity,
            **kwargs
        )
        self.project = os.getenv('GOOGLE_CLOUD_PROJECT')
        self.channels = {}
        self.groups = {}
        self.group_expiry = group_expiry
        self._publisher_client = None
        self._subscriber_client = None
        self._subscribes = {}
        self._messages = {}
        self.queue_limit = 1000

    # Channel layer API
    extensions = ["groups", "flush"]

    @property
    def publisher_client(self):
        if self._publisher_client is None:
            self._publisher_client = pubsub.PublisherClient()
        return self._publisher_client

    @property
    def subscriber_client(self):
        if self._subscriber_client is None:
            self._subscriber_client = pubsub.SubscriberClient()
        return self._subscriber_client

    async def send(self, channel, message):
        """
        Send a message onto a (general or specific) channel.
        """
        # Typecheck
        assert isinstance(message, dict), "message is not a dict"
        assert self.valid_channel_name(channel), "Channel name not valid"
        # If it's a process-local channel, strip off local part and stick full
        # name in message
        assert "__asgi_channel__" not in message

        if channel in self.channels:
            topic_name = self.channels.get(channel)
        else:
            topic_name = 'projects/{project_id}/topics/{topic}'.format(
                project_id=self.project,
                topic=channel,  # Set this to something appropriate.
            )
            self.publisher_client.create_topic(name=topic_name)
            self.channels[channel] = topic_name
        future = self.publisher_client.publish(topic_name, b'', **message)
        future.result()

    async def receive(self, channel):
        """
        Receive the first message that arrives on the channel.
        If more than one coroutine waits on the same channel, a random one
        of the waiting coroutines will get the result.
        """
        assert self.valid_channel_name(channel)

        if channel in self._subscribes:
            subscription_path = self._subscribes[channel]
        else:
            if channel in self.channels:
                topic_name = self.channels.get(channel)
            else:
                topic_name = 'projects/{project_id}/topics/{topic}'.format(
                    project_id=self.project,
                    topic=channel,  # Set this to something appropriate.
                )
                self.publisher_client.create_topic(name=topic_name)
                self.channels[channel] = topic_name
            sub = await self.generate_sub_name(channel)
            subscription_path = self.subscriber_client.subscription_path(self.project, sub)
            self._subscribes[channel] = subscription_path
            self.subscriber_client.create_subscription(request={"name": subscription_path, "topic": topic_name})
            self.subscriber_client.subscribe(subscription_path, functools.partial(self._receive_callback, channel))

        while channel not in self._messages:
            await asyncio.sleep(1)
        while self._messages[channel].empty():
            await asyncio.sleep(0.1)

        message = self._messages[channel].get()
        message.ack()
        return message.attributes

    def _receive_callback(self, channel, message):
        if channel not in self._messages:
            self._messages[channel] = SetQueue(maxsize=self.queue_limit)
        self._messages[channel].put(message)

    async def generate_sub_name(self, channel):
        """
        Returns a new subsciption name that can be used by sub client for this client specifically
        """
        return "%s.gce_sub.%s" % (
            channel,
            uuid.uuid4(),
        )

    async def new_channel(self, prefix="specific."):
        """
        Returns a new channel name that can be used by something in our
        process as a specific channel.
        """
        return "%s.gce_pubsub_channel.%s" % (
            prefix,
            "".join(random.choice(string.ascii_letters) for i in range(12)),
        )

    # Expire cleanup
    def _clean_expired(self):
        """
        Goes through all messages and groups and removes those that are expired.
        Any channel with an expired message is removed from all groups.
        """
        # Group Expiration
        timeout = int(time.time()) - self.group_expiry
        for group in self.groups:
            for channel in list(self.groups.get(group, set())):
                # If join time is older than group_expiry end the group membership
                if (
                    self.groups[group][channel]
                    and int(self.groups[group][channel]) < timeout
                ):
                    # Delete from group
                    del self.groups[group][channel]

    # Flush extension
    async def flush(self):
        self.channels = {}
        self.groups = {}
        self._messages = {}
        for sub_path in self._subscribes:
            self.subscriber_client.delete_subscription(request={"subscription": sub_path})

    async def close(self):
        # Nothing to go
        pass

    def _remove_from_groups(self, channel):
        """
        Removes a channel from all groups. Used when a message on it expires.
        """
        for channels in self.groups.values():
            if channel in channels:
                del channels[channel]

    # Groups extension
    async def group_add(self, group, channel):
        """
        Adds the channel name to a group.
        """
        # Check the inputs
        assert self.valid_group_name(group), "Group name not valid"
        assert self.valid_channel_name(channel), "Channel name not valid"
        # Add to group dict
        self.groups.setdefault(group, {})
        self.groups[group][channel] = time.time()

    async def group_discard(self, group, channel):
        # Both should be text and valid
        assert self.valid_channel_name(channel), "Invalid channel name"
        assert self.valid_group_name(group), "Invalid group name"
        # Remove from group set
        if group in self.groups:
            if channel in self.groups[group]:
                del self.groups[group][channel]
            if not self.groups[group]:
                del self.groups[group]

    async def group_send(self, group, message):
        # Check types
        assert isinstance(message, dict), "Message is not a dict"
        assert self.valid_group_name(group), "Invalid group name"
        # Run clean
        self._clean_expired()
        # Send to each channel
        for channel in self.groups.get(group, set()):
            try:
                await self.send(channel, message)
            except ChannelFull:
                pass


dj_channels_pubsub/layer.py
import asyncio
import collections
import copy
import functools
import os
import queue
import random
import string
import time
import uuid
from copy import deepcopy

from channels.exceptions import ChannelFull
from channels.layers import BaseChannelLayer

from google.cloud import pubsub
from ordered_set import OrderedSet


class GcePubSubChannelLayer(BaseChannelLayer):
    """
    Google cloud Pub/Sub channel layer implementation
    """

    def __init__(
        self,
        expiry=60,
        group_expiry=86400,
        capacity=100,
        channel_capacity=None,
        **kwargs
    ):
        super().__init__(
            expiry=expiry,
            capacity=capacity,
            channel_capacity=channel_capacity,
            **kwargs
        )
        self.channels = {}
        self.groups = {}
        self.group_expiry = group_expiry
        self._publisher_client = None
        self._subscriber_client = None
        self._subscribes = {}
        self.project = "agenda-test-a84a5"  # os.getenv('GOOGLE_CLOUD_PROJECT')
        self._messages = {}

    # Channel layer API

    extensions = ["groups", "flush"]

    @property
    def publisher_client(self):
        if self._publisher_client is None:
            self._publisher_client = pubsub.PublisherClient()
        return self._publisher_client

    @property
    def subscriber_client(self):
        if self._subscriber_client is None:
            self._subscriber_client = pubsub.SubscriberClient()
        return self._subscriber_client

    async def send(self, channel, message):
        """
        Send a message onto a (general or specific) channel.
        """
        # Typecheck
        assert isinstance(message, dict), "message is not a dict"
        assert self.valid_channel_name(channel), "Channel name not valid"
        # If it's a process-local channel, strip off local part and stick full
        # name in message
        assert "__asgi_channel__" not in message

        if channel in self.channels:
            topic_name = self.channels.get(channel)
        else:
            topic_name = 'projects/{project_id}/topics/{topic}'.format(
                project_id=self.project,
                topic=channel,  # Set this to something appropriate.
            )
            self.publisher_client.create_topic(name=topic_name)
            self.channels[channel] = topic_name
        future = self.publisher_client.publish(topic_name, b'', **message)
        future.result()

    async def receive(self, channel):
        """
        Receive the first message that arrives on the channel.
        If more than one coroutine waits on the same channel, a random one
        of the waiting coroutines will get the result.
        """
        assert self.valid_channel_name(channel)

        if channel in self._subscribes:
            subscription_path = self._subscribes[channel]
        else:
            if channel in self.channels:
                topic_name = self.channels.get(channel)
            else:
                topic_name = 'projects/{project_id}/topics/{topic}'.format(
                    project_id=self.project,
                    topic=channel,  # Set this to something appropriate.
                )
                self.publisher_client.create_topic(name=topic_name)
                self.channels[channel] = topic_name
            sub = await self.generate_sub_name(channel)
            subscription_path = self.subscriber_client.subscription_path(self.project, sub)
            self._subscribes[channel] = subscription_path
            self.subscriber_client.create_subscription(request={"name": subscription_path, "topic": topic_name})
            self.subscriber_client.subscribe(subscription_path, functools.partial(self._receive_callback, channel))

        while channel not in self._messages:
            await asyncio.sleep(1)
        while len(self._messages[channel]) == 0:
            await asyncio.sleep(0.1)
        # if channel in self._messages:
        #     if self._messages[channel].qsize() > 0:
        message = self._messages[channel][0]
        message.ack()
        self._messages[channel].discard(message)
        return message.attributes

    def _receive_callback(self, channel, message):
        print("message")
        print(channel)
        print(message)
        if channel not in self._messages:
            self._messages[channel] = OrderedSet()
        self._messages[channel].add(message)

    async def generate_sub_name(self, channel):
        """
        Returns a new subsciption name that can be used by sub client for this client specifically
        """
        return "%s.gce_sub.%s" % (
            channel,
            uuid.uuid4(),
        )

    async def new_channel(self, prefix="specific."):
        """
        Returns a new channel name that can be used by something in our
        process as a specific channel.
        """
        return "%s.gce_pubsub_channel.%s" % (
            prefix,
            "".join(random.choice(string.ascii_letters) for i in range(12)),
        )

    # Expire cleanup
    def _clean_expired(self):
        """
        Goes through all messages and groups and removes those that are expired.
        Any channel with an expired message is removed from all groups.
        """
        # Channel cleanup
        for channel, queue in list(self.channels.items()):
            # See if it's expired
            while not queue.empty() and queue._queue[0][0] < time.time():
                queue.get_nowait()
                # Any removal prompts group discard
                self._remove_from_groups(channel)
                # Is the channel now empty and needs deleting?
                if queue.empty():
                    del self.channels[channel]

        # Group Expiration
        timeout = int(time.time()) - self.group_expiry
        for group in self.groups:
            for channel in list(self.groups.get(group, set())):
                # If join time is older than group_expiry end the group membership
                if (
                    self.groups[group][channel]
                    and int(self.groups[group][channel]) < timeout
                ):
                    # Delete from group
                    del self.groups[group][channel]

    # Flush extension
    async def flush(self):
        self.channels = {}
        self.groups = {}
        # TODO close topics and subs

    async def close(self):
        # Nothing to go
        pass

    def _remove_from_groups(self, channel):
        """
        Removes a channel from all groups. Used when a message on it expires.
        """
        for channels in self.groups.values():
            if channel in channels:
                del channels[channel]

    # Groups extension

    async def group_add(self, group, channel):
        """
        Adds the channel name to a group.
        """
        # Check the inputs
        assert self.valid_group_name(group), "Group name not valid"
        assert self.valid_channel_name(channel), "Channel name not valid"
        # Add to group dict
        self.groups.setdefault(group, {})
        self.groups[group][channel] = time.time()

    async def group_discard(self, group, channel):
        # Both should be text and valid
        assert self.valid_channel_name(channel), "Invalid channel name"
        assert self.valid_group_name(group), "Invalid group name"
        # Remove from group set
        if group in self.groups:
            if channel in self.groups[group]:
                del self.groups[group][channel]
            if not self.groups[group]:
                del self.groups[group]

    async def group_send(self, group, message):
        # Check types
        assert isinstance(message, dict), "Message is not a dict"
        assert self.valid_group_name(group), "Invalid group name"
        # Run clean
        # self._clean_expired()
        # Send to each channel
        for channel in self.groups.get(group, set()):
            try:
                print(message)
                await self.send(channel, message)
            except ChannelFull:
                pass
